src/Database/Dpi/Internal.chs

```diff
@@ -624,111 +624,106 @@ newtype Data = Data (NativeTypeNum -> OracleTypeNum -> IO DataValue)
+      NativeTypeInt64      -> fmap (gInt  o . ft)         $ libDataGetInt64 p
+      NativeTypeUint64     -> fmap (gUnt o . ft)          $ libDataGetUint64 p
+      NativeTypeFloat      -> fmap DataFloat              $ libDataGetFloat p
+      NativeTypeDouble     -> fmap DataDouble             $ libDataGetDouble p
+      NativeTypeBytes      -> fmap (gByt o) $ do
         ptr'     <- {#get Data -> value.asBytes.ptr      #} p
         len      <- {#get Data -> value.asBytes.length   #} p
         encoding <- {#get Data -> value.asBytes.encoding #} p >>= ts
         let bytes = (ptr', fromIntegral len)
-        return Data_Bytes{..}
-      go p NativeTypeTimestamp  o = do
-        year           <- {#get Data -> value.asTimestamp.year           #} p
-        month          <- {#get Data -> value.asTimestamp.month          #} p
-        day            <- {#get Data -> value.asTimestamp.day            #} p
-        hour           <- {#get Data -> value.asTimestamp.hour           #} p
-        minute         <- {#get Data -> value.asTimestamp.minute         #} p
-        second         <- {#get Data -> value.asTimestamp.second         #} p
-        fsecond        <- {#get Data -> value.asTimestamp.fsecond        #} p
-        tzHourOffset   <- {#get Data -> value.asTimestamp.tzHourOffset   #} p
-        tzMinuteOffset <- {#get Data -> value.asTimestamp.tzMinuteOffset #} p
-        return $ toTime o Data_Timestamp{..}
-      go p NativeTypeIntervalDs _ = DataIntervalDs <$> do
-        days     <- {#get Data -> value.asIntervalDS.days     #} p
-        hours    <- {#get Data -> value.asIntervalDS.hours    #} p
-        minutes  <- {#get Data -> value.asIntervalDS.minutes  #} p
-        seconds  <- {#get Data -> value.asIntervalDS.seconds  #} p
-        fseconds <- {#get Data -> value.asIntervalDS.fseconds #} p
-        return $ Data_IntervalDS{..}
-      go p NativeTypeIntervalYm _ = DataIntervalYm <$> do
-        years    <- {#get Data -> value.asIntervalYM.years    #} p
-        months   <- {#get Data -> value.asIntervalYM.months   #} p
-        return Data_IntervalYM {..}
-      go p NativeTypeLob        o = gLob o     <$> {#get Data -> value.asLOB    #} p
-      go p NativeTypeObject     _ = DataObject <$> {#get Data -> value.asObject #} p
-      go p NativeTypeStmt       _ = DataStmt   <$> {#get Data -> value.asStmt   #} p
-      go p NativeTypeRowid      _ = DataRowid  <$> {#get Data -> value.asRowid  #} p
-
+        pure $ Data_Bytes{..}
+      NativeTypeTimestamp  -> do
+        p' <- libDataGetTimestamp p
+        if p' == nullPtr
+        then pure $ DataNull t
+        else fmap (toTime o) $ peek p'
+      NativeTypeIntervalDs -> do
+        p' <- libDataGetIntervalDS p
+        if p' == nullPtr
+        then pure $ DataNull t
+        else fmap DataIntervalDs $ peek p'
+      NativeTypeIntervalYm -> do
+        p' <- libDataGetIntervalYM p
+        if p' == nullPtr
```

Comment:
some of the lib* functions return a pointer, so we check if it's null, and map to `DataNull` in those cases